nixos/modules/services/system/self-deploy.nix

```diff
@@ -0,0 +1,174 @@
+  outPath = "${workingDirectory}/system";
+
+  gitWithRepo = "git -C ${repositoryDirectory}";
+
+  renderNixArgs = args:
+    let
+      toArg = key: value:
+        if builtins.isString value
+        then " --argstr ${lib.escapeShellArg key} ${lib.escapeShellArg value}"
+        else " --arg ${lib.escapeShellArg key} ${lib.escapeShellArg (toString value)}";
+    in
+      lib.concatStrings (lib.mapAttrsToList toArg args);
+
+  isPathType = x: lib.strings.isCoercibleToString x && builtins.substring 0 1 (toString x) == "/";
+
+in {
+  options.services.self-deploy = {
+    enable = lib.mkEnableOption "self-deploy";
+
+    nixFile = lib.mkOption {
+      type = lib.types.path;
+
+      default = "/default.nix";
+
+      description = ''
+      Path to nix file in repository. Leading '/' refers to root of
+      git repository.
+      '';
+    };
+
+    nixAttribute = lib.mkOption {
+      type = lib.types.str;
+
+      description = ''
+      Attribute of `nixFile` that builds the current system.
+      '';
+    };
+
+    nixArgs = lib.mkOption {
+      type = lib.types.attrs;
+
+      default = {};
+
+      description = ''
+      Arguments to `nix-build` passed as `--argstr` or `--arg` depending on
+      the type.
+      '';
+    };
+
+    switchCommand = lib.mkOption {
+      type = lib.types.enum [ "boot" "switch" "dry-activate" "test" ];
+
+      default = "switch";
+
+      description = ''
+      The `switch-to-configuration` subcommand used.
+      '';
+    };
+
+    repository = lib.mkOption {
+      type = with lib.types; oneOf [ path str ];
+
+      description = ''
+      The repository to fetch from. Must be properly formatted for git.
+
+      If this value is set to a path (must begin with `/`) then it's
+      assumed that the repository is local and the resulting service
+      won't wait for the network to be up.
+
+      If the repository will be fetched over SSH, you must add an
+      entry to `programs.ssh.knownHosts` for the SSH host for the fetch
+      to be successful.
+      '';
+    };
+
+    sshKeyFile = lib.mkOption {
+      type = with lib.types; nullOr path;
+
+      default = null;
+
+      description = ''
+      Path to SSH private key used to fetch private repositories over
+      SSH.
+      '';
+    };
+
+    branch = lib.mkOption {
+      type = lib.types.str;
+
+      default = "master";
+
+      description = ''
+      Branch to track
+
+      You can also specify a revision or tag, too, if you want to pin this
+      machine to a specific commit.
+      '';
+    };
+
+    startAt = lib.mkOption {
+      type = with lib.types; either str (listOf str);
+
+      default = "hourly";
+
+      description = ''
+      The schedule on which to run the `self-deploy` service. Format
+      specified by `systemd.time 7`.
+
+      This value can also be a list of `systemd.time 7` formatted
+      strings, in which case the service will be started on multiple
+      schedules.
```

Comment:
I think `systemd.time(7)` with the parenthesis is more correct, or you could say `man 7 systemd.time` if you want to be more helpful.

Reply:
I was just matching the wording used in [`systemd.services.<name>.startAt`](https://search.nixos.org/options?channel=20.09&show=systemd.services.%3Cname%3E.startAt&from=0&size=50&sort=relevance&query=systemd.services.%3Cname%3E.startAt) since this value is just copied directly into it.

Reply:
Ah okay, then disregard my comment 👍